golem-cli/src/diagnose.rs

```diff
@@ -0,0 +1,893 @@
+            GuestLanguage::Go => Some(Language::Go),
+            GuestLanguage::C => Some(Language::CCcp),
+            GuestLanguage::Zig => Some(Language::Zig),
+            GuestLanguage::JavaScript => Some(Language::JsTs),
+            GuestLanguage::TypeScript => Some(Language::JsTs),
+            GuestLanguage::CSharp => None,
+            GuestLanguage::Swift => None,
+            GuestLanguage::Grain => None,
+            GuestLanguage::Python => Some(Language::Python),
+            GuestLanguage::Scala2 => None,
+        };
+
+        language.map(|language| SelectedLanguage {
+            language,
+            project_dir: PathBuf::from("."),
+            detected_by_reason: None,
+        })
+    }
+
+    pub fn from_env() -> Option<SelectedLanguage> {
+        let ordered_language_project_hint_files: Vec<(&str, Language)> = vec![
+            ("build.zig", Language::Zig),
+            ("Cargo.toml", Language::Rust),
+            ("go.mod", Language::Go),
+            ("package.json", Language::JsTs),
+            ("Makefile", Language::CCcp),
+            ("main.py", Language::Python),
+        ];
+
+        let detect_in_dir = |dir: &Path| -> Option<SelectedLanguage> {
+            ordered_language_project_hint_files
+                .iter()
+                .find_map(|(file, language)| {
+                    let path = dir.join(Path::new(file));
+                    path.exists().then_some(SelectedLanguage {
+                        language: *language,
+                        project_dir: path
+                            .parent()
+                            .expect("Failed to get parent for project file")
+                            .to_path_buf(),
+                        detected_by_reason: Some(format!(
+                            "Detected project file: {}",
+                            path.to_string_lossy()
+                        )),
+                    })
+                })
+        };
+
+        fn is_dir(entry: &DirEntry) -> bool {
+            entry
+                .metadata()
+                .expect("Failed to get dir entry metadata")
+                .is_dir()
+        }
+
+        fn is_hidden(entry: &DirEntry) -> bool {
+            let file_name = entry
+                .file_name()
+                .to_str()
+                .expect("Failed to get file name from dir entry");
+            file_name != "." && file_name.starts_with(".")
+        }
+
+        // Searching - down
+        {
+            let language = walkdir::WalkDir::new(".")
+                .max_depth(4)
+                .into_iter()
+                .filter_entry(|e| is_dir(e) && !is_hidden(e))
+                .find_map(|e| e.ok().and_then(|e| detect_in_dir(e.path())));
+
+            if language.is_some() {
+                return language;
+            }
+        }
+
+        // Searching - up
+        {
+            let starting_dir = std::env::current_dir().expect("Failed to get current dir");
```

Comment:
Trying to understand things here. 

This is for being able to run diagnose command from anywhere in the project ?

Reply:
i wanted to make it work for our current guides / examples, and there are 3 main kind of projects currently:
- simple ones with only one component,
- multi components examples with a root project
- lower level doc examples with a project folder, where every subfolder is a separate project

also, i wanted to make it work in case you are inside of some project src folder, for some examples we have those, for some not.

so in the end it search in both directions for hints